README
NAME
    Test::Builder::Mock::Class - Simulating other classes for Test::Builder

SYNOPSIS
      use Test::Builder::Mock::Class ':all';

      require 'Net::FTP::Mock';
      mock_class 'Net::FTP' => 'Net::FTP::Mock';
      my $mock_object = Net::FTP::Mock->new;

      # anonymous mocked class
      my $metamock = mock_anon_class 'Net::FTP';
      my $mock_object = $metamock->new_object;

      # anonymous class with role applied
      my $metamock = Test::Builder::Mock::Class->create_anon_class(
          roles => [ 'My::Handler::Role' ],
      );
      my $mock_object = $metamock->new_object;

DESCRIPTION
    This module adds support for standard Test::Builder framework
    (Test::Simple or Test::More) to Test::Mock::Class.

    Mock class can be used to create mock objects which can simulate the
    behavior of complex, real (non-mock) objects and are therefore useful
    when a real object is impractical or impossible to incorporate into a
    unit test.

    See Test::Mock::Class for more detailed documentation.

BUGS
    The API is not stable yet and can be changed in future.

AUTHOR
    Piotr Roszatycki <dexter@cpan.org>

LICENSE
    Based on SimpleTest, an open source unit test framework for the PHP
    programming language, created by Marcus Baker, Jason Sweat, Travis
    Swicegood, Perrick Penet and Edward Z. Yang.

    Copyright (c) 2009, 2010 Piotr Roszatycki <dexter@cpan.org>.

    This program is free software; you can redistribute it and/or modify it
    under GNU Lesser General Public License.



Changes
------------------------------------------------------------------------
0.0203 | Piotr Roszatycki <dexter@cpan.org> | 2010-05-28

------------------------------------------------------------------------
0.0202 | Piotr Roszatycki <dexter@cpan.org> | 2010-05-14

Fixes:

* Class::MOP requires IO::* modules explicitly.  Implicit load does not work
  correctly.

Changes:

* Requires Test::Mock::Class >= 0.0303.

------------------------------------------------------------------------
0.0201 | Piotr Roszatycki <dexter@cpan.org> | 2009-10-28

Fixes:

* Compatibility with Moose >= 0.89.

Changes:

* Test::Builder::Mock::Class::Role::Object:
  - Use new Moose "with" keyword syntax.

------------------------------------------------------------------------
0.02 | Piotr Roszatycki <dexter@cpan.org> | 2009-04-18

Changes:

* Function "mock_anon_class" has optional argument.

------------------------------------------------------------------------
0.0102 | Piotr Roszatycki <dexter@cpan.org> | 2009-04-16

Fixes:

* This module is under LGPL only.

------------------------------------------------------------------------
0.0101 | Piotr Roszatycki <dexter@cpan.org> | 2009-04-15

Changes:

* Aliased methods from inherited roles are used.

------------------------------------------------------------------------
0.01 | Piotr Roszatycki <dexter@cpan.org> | 2009-04-14

* First CPAN release.

------------------------------------------------------------------------


lib/Test/Builder/Mock/Class.pm
#!/usr/bin/perl -c

package Test::Builder::Mock::Class;

=head1 NAME

Test::Builder::Mock::Class - Simulating other classes for Test::Builder

=head1 SYNOPSIS

  use Test::Builder::Mock::Class ':all';
  use Test::More 'no_plan';

  # concrete mock class
  mock_class 'Net::FTP' => 'Net::FTP::Mock';
  my $mock_object1 = Net::FTP::Mock->new;
  $mock_object1->mock_tally;

  # anonymous mocked class
  my $metamock2 = mock_anon_class 'Net::FTP';
  my $mock_object2 = $metamock2->new_object;
  $mock_object2->mock_tally;

  # anonymous class with role applied
  my $metamock3 = Test::Builder::Mock::Class->create_mock_anon_class(
      class => 'Net::FTP',
      roles => [ 'My::Handler::Role' ],
  );
  my $mock_object3 = $metamock3->new_object;
  $mock_object3->mock_tally;

=head1 DESCRIPTION

This module adds support for standard L<Test::Builder> framework
(L<Test::Simple> or L<Test::More>) to L<Test::Mock::Class>.

Mock class can be used to create mock objects which can simulate the behavior
of complex, real (non-mock) objects and are therefore useful when a real
object is impractical or impossible to incorporate into a unit test.

See L<Test::Mock::Class> for more detailed documentation.

=for readme stop

=cut

use 5.006;

use strict;
use warnings;

our $VERSION = '0.0203';

use Moose;



=head1 INHERITANCE

=over 2

=item *

extends L<Moose::Meta::Class>

=cut

extends 'Moose::Meta::Class';

=item *

with L<Test::Builder::Mock::Class::Role::Meta::Class>

=back

=cut

with 'Test::Builder::Mock::Class::Role::Meta::Class';


use namespace::clean -except => 'meta';


=head1 FUNCTIONS

=over

=cut

BEGIN {
    my %exports = ();

=item B<mock_class>( I<class> : Str, I<mock_class> : Str = undef ) : Moose::Meta::Class

Creates the concrete mock class based on original I<class>.  If the name of
I<mock_class> is undefined, its name is created based on name of original
I<class> with added C<::Mock> suffix.

The function returns the metaclass object of new I<mock_class>.

=cut

    $exports{mock_class} = sub {
        sub ($;$) {
            return __PACKAGE__->create_mock_class(
                defined $_[1] ? $_[1] : $_[0] . '::Mock',
                class => $_[0],
            );
        };
    };

=item B<mock_anon_class>( I<class> : Str = undef ) : Moose::Meta::Class

Creates an anonymous mock class based on original I<class>.  The name of this
class is automatically generated.  If I<class> argument not defined, the empty
mock class is created.

The function returns the metaobject of new mock class.

=back

=cut

    $exports{mock_anon_class} = sub {
        sub (;$) {
            return __PACKAGE__->create_mock_anon_class(
                defined $_[0] ? (class => $_[0]) : (),
            );
        };
    };

=head1 IMPORTS

=over

=cut

    my %groups = ();

=item Test::Builder::Mock::Class ':all';

Imports all functions into caller's namespace.

=back

=cut

    $groups{all} = [ keys %exports ];

    require Sub::Exporter;
    Sub::Exporter->import(
        -setup => {
            exports => [ %exports ],
            groups => \%groups,
        },
    );
};


1;


=begin umlwiki

= Class Diagram =

[                           Test::Builder::Mock::Class
 ------------------------------------------------------------------------------------
 ------------------------------------------------------------------------------------
 <<utility>> mock_class( class : Str, mock_class : Str = undef ) : Moose::Meta::Class
 <<utility>> mock_anon_class( class : Str ) : Moose::Meta::Class
                                                                                     ]

[Test::Builder::Mock::Class] ---|> [Moose::Meta::Class] [<<role>> Test::Builder::Mock::Class::Role::Meta::Class]

=end umlwiki

=head1 EXAMPLE

The C<Test::Builder::Mock::Class> fits perfectly to L<Test::Builder>
(L<Test::Simple> or L<Test::More>) tests. It adds automatically the tests for
each C<mock_invoke> (which is called implicitly by all mock methods) and
C<mock_tally>.  It means that you need to add these tests to your test plan.

Example code:

  package My::ExampleTest;

  use Test::More 'no_plan';
  use Test::Builder::Mock::Class ':all';

  require 'IO::File';
  my $mock = mock_anon_class 'IO::File';
  my $io = $mock->new_object;
  $io->mock_return( open => 1, args => [qr//, 'r'] );

  ok( $io->open('/etc/passwd', 'r'), '$io->open' );
  $io->mock_tally;

=head1 BUGS

L<Moose> after version 1.05 calls C<BUILDALL> method automatically, so this
is one more test to whole plan.

L<Test::More> needs an exact count of all tests and it will be different for
L<Moose> before and after version 1.05.  It can be overriden with following
ways:

  # No plan at all
  use Test::More 'no_plan';

  # Different plans depend on Moose version
  use Test::More;
  require Moose;
  plan tests => (Moose->VERSION >= 1.05 ? 10 : 8);

  # Require Moose >= 1.05
  use Moose 1.05 ();
  use Test::More tests => 10;

=head1 SEE ALSO

Mock metaclass API: L<Test::Builder::Mock::Class::Role::Meta::Class>,
L<Moose::Meta::Class>.

Mock object methods: L<Test::Builder::Mock::Class::Role::Object>.

Perl standard testing: L<Test::Builder>, L<Test::Simple>, L<Test::More>.

Mock classes for xUnit-like testing (L<Test::Unit::Lite>):
L<Test::Mock::Class>.

Other implementations: L<Test::MockObject>, L<Test::MockClass>.

=for readme continue

=head1 BUGS

The API is not stable yet and can be changed in future.

=head1 AUTHOR

Piotr Roszatycki <dexter@cpan.org>

=head1 LICENSE

Based on SimpleTest, an open source unit test framework for the PHP
programming language, created by Marcus Baker, Jason Sweat, Travis Swicegood,
Perrick Penet and Edward Z. Yang.

Copyright (c) 2009, 2010 Piotr Roszatycki <dexter@cpan.org>.

This program is free software; you can redistribute it and/or modify it
under GNU Lesser General Public License.


lib/Test/Builder/Mock/Class/Role/Object.pm
#!/usr/bin/perl -c

package Test::Builder::Mock::Class::Role::Object;

=head1 NAME

Test::Builder::Mock::Class::Role::Object - Role for base object of mock class

=head1 DESCRIPTION

This role provides an API for changing behavior of mock class.

=cut

use 5.006;

use strict;
use warnings;

our $VERSION = '0.0202';

use Moose::Role 0.89;


=head1 INHERITANCE

=over 2

=item *

with L<Test::Mock::Class::Role::Object>

=back

=cut

with 'Test::Mock::Class::Role::Object' => {
    -alias => {
        mock_invoke => '_mock_invoke_base',
        mock_tally  => '_mock_tally_base',
    },
    -excludes => [ 'mock_invoke', 'mock_tally' ],
};


use English '-no_match_vars';

use Test::Builder;


use Exception::Base (
    '+ignore_package' => [__PACKAGE__],
);


=head1 ATTRIBUTES

=over

=item B<_mock_test_builder> : Test::Builder

The L<Test::Builder> singleton object.

=back

=cut

has '_mock_test_builder' => (
    is      => 'rw',
    default => sub { Test::Builder->new },
);


use namespace::clean -except => 'meta';


## no critic qw(RequireCheckingReturnValueOfEval)

=head1 METHODS

=over

=item B<mock_tally>(I<>) : Self

Check the expectations at the end.  See L<Test::Mock::Class::Role::Object> for
more description.

The test passes if original C<mock_tally> method doesn't throw an exception.

=cut

sub mock_tally {
    my ($self) = @_;

    my $return = eval {
        $self->_mock_tally_base;
    };
    $self->_mock_test_builder->is_eq($EVAL_ERROR, '', 'mock_tally()');

    return $return;
};


=item B<mock_invoke>( I<method> : Str, I<args> : Array ) : Any

Returns the expected value for the method name and checks expectations.  See
L<Test::Mock::Class::Role::Object> for more description.

The test passes if original C<mock_tally> method doesn't throw an exception.

=cut

sub mock_invoke {
    my ($self, $method, @args) = @_;

    my ($return, @return);
    eval {
        if (wantarray) {
            @return = $self->_mock_invoke_base($method, @args);
        }
        else {
            $return = $self->_mock_invoke_base($method, @args);
        };
    };
    $self->_mock_test_builder->is_eq($EVAL_ERROR, '', "mock_invoke($method)");

    return wantarray ? @return : $return;
};


1;


=back

=begin umlwiki

= Class Diagram =

[                                <<role>>
                    Test::Builder::Mock::Class::Role::Object
 -----------------------------------------------------------------------------
 #_mock_test_builder : Test::Builder
 -----------------------------------------------------------------------------
 +mock_tally() : Self
 +mock_invoke( method : Str, args : Array ) : Any
                                                                              ]

[Test::Builder::Mock::Class::Role::Object] ---|> [<<role>> Test::Mock::Class::Role::Object]

=end umlwiki

=head1 SEE ALSO

L<Test::Mock::Class::Role::Object>, L<Test::Mock::Class>.

=head1 BUGS

The API is not stable yet and can be changed in future.

=head1 AUTHOR

Piotr Roszatycki <dexter@cpan.org>

=head1 LICENSE

Based on SimpleTest, an open source unit test framework for the PHP
programming language, created by Marcus Baker, Jason Sweat, Travis Swicegood,
Perrick Penet and Edward Z. Yang.

Copyright (c) 2009, 2010 Piotr Roszatycki <dexter@cpan.org>.

This program is free software; you can redistribute it and/or modify it
under GNU Lesser General Public License.


lib/Test/Builder/Mock/Class/Role/Meta/Class.pm
#!/usr/bin/perl -c

package Test::Builder::Mock::Class::Role::Meta::Class;

=head1 NAME

Test::Builder::Mock::Class::Role::Meta::Class - Metaclass for mock class

=head1 DESCRIPTION

This role provides an API for defining mock class.

=cut

use 5.006;

use strict;
use warnings;

our $VERSION = '0.0202';

use Moose::Role;


=head1 INHERITANCE

=over 2

=item *

L<Test::Mock::Class::Role::Meta::Class>

=back

=cut

with 'Test::Mock::Class::Role::Meta::Class';


use Exception::Base (
    '+ignore_package' => [__PACKAGE__],
);


=head1 ATTRIBUTES

=over

=item B<mock_base_object_role> : Str = "Test::Builder::Mock::Class::Role::Object"

Base object role for mock class.  The default is
L<Test::Builder::Mock::Class::Role::Object>.

=cut

has 'mock_base_object_role' => (
    is      => 'rw',
    default => 'Test::Builder::Mock::Class::Role::Object',
);


use namespace::clean -except => 'meta';


1;


=back

=begin umlwiki

= Class Diagram =

[                              <<role>>
                  Test::Builder::Mock::Class::Role::Meta::Class
 ------------------------------------------------------------------------------
 +mock_base_object_role = "Test::Builder::Mock::Class::Role::Object"
 ------------------------------------------------------------------------------
                                                                               ]

[Test::Builder::Mock::Class::Role::Meta::Class] ---|> [<<role>> Test::Mock::Class::Role::Meta::Class]

=end umlwiki

=head1 SEE ALSO

L<Test::Mock::Class::Role::Meta::Class>, L<Test::Builder::Mock::Class>.

=head1 BUGS

The API is not stable yet and can be changed in future.

=head1 AUTHOR

Piotr Roszatycki <dexter@cpan.org>

=head1 LICENSE

Based on SimpleTest, an open source unit test framework for the PHP
programming language, created by Marcus Baker, Jason Sweat, Travis Swicegood,
Perrick Penet and Edward Z. Yang.

Copyright (c) 2009, 2010 Piotr Roszatycki <dexter@cpan.org>.

This program is free software; you can redistribute it and/or modify it
under GNU Lesser General Public License.


Build.PL
#!/usr/bin/perl

use 5.006;

use strict;
use warnings;

use Module::Build;

my $module = 'Test::Builder::Mock::Class';

(my $dist = $module) =~ s/::/-/g;

my $build = Module::Build->new(
    module_name => $module,
    license => 'lgpl',
    requires => {
        'perl'               => 5.006,
        'namespace::clean'   => 0,
        'Moose'              => 0.89,
        'Test::Mock::Class'  => 0.0201,
        'Test::Builder'      => 0,
    },
    optional => {
    },
    build_requires => {
    },
    auto_configure_requires => 0,
    create_makefile_pl => 'traditional',
    create_readme => 1,
    create_license => 1,
    sign => 1,
    recursive_test_files => 1,
    add_to_cleanup => [
        '*.bak', '*.gz', '*.tgz', '*.ppd', '*.ppmx', '*.deb', '*.bat',
        'Makefile', 'pm_to_blib', 'cover_db', 'xt/Debian_CPANTS.txt',
        "$dist-*",
    ],
    meta_merge => {
        resources => {
            bugtracker => "http://rt.cpan.org/NoAuth/Bugs.html?Dist=$dist",
            repository => "http://dexter-cpan.googlecode.com/svn/$dist",
        },
    },
);

$build->create_build_script;
